File: app/src/main/java/org/amoustakos/boilerplate/util/network/MobileDataUtil.java
package org.amoustakos.boilerplate.util.network;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.telephony.TelephonyManager;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import timber.log.Timber;

import static org.amoustakos.boilerplate.util.network.NetworkUtil.getNetworkInfo;


/**
 * NOTE: NEED TO INSTALL AS SYSTEM APP FOR THIS TO WORK.
 */

public class MobileDataUtil {


    /**
     * Checks if the device is connected to a mobile network
     */
    public static boolean isConnected(Context context){
        final ConnectivityManager conManager = (ConnectivityManager)
                context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo net = conManager.getActiveNetworkInfo();

        return net != null
                && net.isConnected()
                && ConnectivityManager.TYPE_MOBILE == net.getType();
    }


	/**
	 * Retrieves the state of the mobile data (on = true, off = false)
	 */
	public static boolean getState(Context context){
        try{
            TelephonyManager telephonyService = (TelephonyManager)
                    context.getSystemService(Context.TELEPHONY_SERVICE);
            Method getMobileDataEnabledMethod =
                    telephonyService.getClass().getDeclaredMethod("getDataEnabled");

            if (null != getMobileDataEnabledMethod){
                return (Boolean) getMobileDataEnabledMethod.invoke(telephonyService);
            }
        }catch (Exception e){
            Timber.w(e, e.getMessage());
        }
        return false;
    }


	/**
	 * Sets the state of the mobile data (on = true, off = false)
	 */
	public static boolean setState(Context context, boolean active){
        final ConnectivityManager conman = (ConnectivityManager)
                context.getSystemService(Context.CONNECTIVITY_SERVICE);
        Method dataMtd;
        try {
            dataMtd =ConnectivityManager.class.getDeclaredMethod("setMobileDataEnabled", boolean.class);
        } catch (NoSuchMethodException e) {
            Timber.w(e, e.getMessage());
            return false;
        }
        dataMtd.setAccessible(true);
        try {
            dataMtd.invoke(conman, active);
        } catch (IllegalAccessException | InvocationTargetException e) {
            Timber.w(e, e.getMessage());
            return false;
        }
        return true;
    }


    /**
     * Checks if there is fast connectivity
     */
    public static boolean isConnectedFast(Context context) {
        NetworkInfo info = getNetworkInfo(context);
        return info != null && info.isConnected() && isConnectionFast(info.getType(), info.getSubtype());
    }


    /**
     * Checks if the connection is fast
     */
    public static boolean isConnectionFast(int type, int subType) {
        if (type == ConnectivityManager.TYPE_WIFI)
            return true;
        else if (type == ConnectivityManager.TYPE_MOBILE)
            switch (subType) {
                case TelephonyManager.NETWORK_TYPE_1xRTT:
                    return false; // ~ 50-100 kbps
                case TelephonyManager.NETWORK_TYPE_CDMA:
                    return false; // ~ 14-64 kbps
                case TelephonyManager.NETWORK_TYPE_EDGE:
                    return false; // ~ 50-100 kbps
                case TelephonyManager.NETWORK_TYPE_EVDO_0:
                    return true; // ~ 400-1000 kbps
                case TelephonyManager.NETWORK_TYPE_EVDO_A:
                    return true; // ~ 600-1400 kbps
                case TelephonyManager.NETWORK_TYPE_GPRS:
                    return false; // ~ 100 kbps
                case TelephonyManager.NETWORK_TYPE_HSDPA:
                    return true; // ~ 2-14 Mbps
                case TelephonyManager.NETWORK_TYPE_HSPA:
                    return true; // ~ 700-1700 kbps
                case TelephonyManager.NETWORK_TYPE_HSUPA:
                    return true; // ~ 1-23 Mbps
                case TelephonyManager.NETWORK_TYPE_UMTS:
                    return true; // ~ 400-7000 kbps
                case TelephonyManager.NETWORK_TYPE_EHRPD:
                    return true; // ~ 1-2 Mbps
                case TelephonyManager.NETWORK_TYPE_EVDO_B:
                    return true; // ~ 5 Mbps
                case TelephonyManager.NETWORK_TYPE_HSPAP:
                    return true; // ~ 10-20 Mbps
                case TelephonyManager.NETWORK_TYPE_IDEN:
                    return false; // ~25 kbps
                case TelephonyManager.NETWORK_TYPE_LTE:
                    return true; // ~ 10+ Mbps
                // Unknown
//                case TelephonyManager.NETWORK_TYPE_UNKNOWN:
                default:
                    return false;
            }
        else
            return false;
    }
}

File: utils_network/src/main/java/org/amoustakos/utils/network/MobileDataUtil.kt
@file:Suppress("DEPRECATION")
package org.amoustakos.utils.network

import android.annotation.SuppressLint
import android.content.Context
import android.net.ConnectivityManager
import android.telephony.TelephonyManager
import timber.log.Timber
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method


@SuppressLint("PrivateApi")
@Deprecated("To be replaced with high/low bandwidth utility")
object MobileDataUtil {

    /**
     * Checks if the device is connected to a mobile network
     */
    @Throws(NullPointerException::class)
    fun isConnected(context: Context): Boolean {
        val conManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val net = conManager.activeNetworkInfo

        return (net != null
                && net.isConnected
                && ConnectivityManager.TYPE_MOBILE == net.type)
    }


    /**
     * Retrieves the state of the mobile data (on = true, off = false)
     */
    fun getState(context: Context): Boolean {
        try {
            val telephonyService =
                    context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
            return telephonyService.javaClass.getDeclaredMethod("getDataEnabled")
		            .invoke(telephonyService) as Boolean
        } catch (e: Exception) {
            Timber.w(e)
        }

        return false
    }


    /**
     * Sets the state of the mobile data (on = true, off = false)
     */
    fun setState(context: Context, active: Boolean): Boolean {
        val conman = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val dataMtd: Method
        try {
            dataMtd = ConnectivityManager::class.java.
                    getDeclaredMethod("setMobileDataEnabled",
                            Boolean::class.javaPrimitiveType)
        } catch (e: NoSuchMethodException) {
            Timber.w(e)
            return false
        }

        dataMtd.isAccessible = true
        try {
            dataMtd.invoke(conman, active)
        } catch (e: IllegalAccessException) {
            Timber.w(e)
            return false
        } catch (e: InvocationTargetException) {
            Timber.w(e)
            return false
        }

        return true
    }


    /**
     * Checks if there is fast connectivity
     */
    fun isConnectedFast(context: Context): Boolean {
        val info = NetworkUtil.getNetworkInfo(context)
        return info != null && info.isConnected && isConnectionFast(info.type, info.subtype)
    }


    /**
     * Checks if the connection is fast
     */
    fun isConnectionFast(type: Int, subType: Int): Boolean {
        return when (type) {
		        ConnectivityManager.TYPE_WIFI           -> true
		        ConnectivityManager.TYPE_MOBILE         -> when (subType) {
			        TelephonyManager.NETWORK_TYPE_1xRTT     -> false // ~ 50-100 kbps
			        TelephonyManager.NETWORK_TYPE_CDMA      -> false // ~ 14-64 kbps
			        TelephonyManager.NETWORK_TYPE_EDGE      -> false // ~ 50-100 kbps
			        TelephonyManager.NETWORK_TYPE_EVDO_0    -> true // ~ 400-1000 kbps
			        TelephonyManager.NETWORK_TYPE_EVDO_A    -> true // ~ 600-1400 kbps
			        TelephonyManager.NETWORK_TYPE_GPRS      -> false // ~ 100 kbps
			        TelephonyManager.NETWORK_TYPE_HSDPA     -> true // ~ 2-14 Mbps
			        TelephonyManager.NETWORK_TYPE_HSPA      -> true // ~ 700-1700 kbps
			        TelephonyManager.NETWORK_TYPE_HSUPA     -> true // ~ 1-23 Mbps
			        TelephonyManager.NETWORK_TYPE_UMTS      -> true // ~ 400-7000 kbps
			        TelephonyManager.NETWORK_TYPE_EHRPD     -> true // ~ 1-2 Mbps
			        TelephonyManager.NETWORK_TYPE_EVDO_B    -> true // ~ 5 Mbps
			        TelephonyManager.NETWORK_TYPE_HSPAP     -> true // ~ 10-20 Mbps
			        TelephonyManager.NETWORK_TYPE_IDEN      -> false // ~25 kbps
			        TelephonyManager.NETWORK_TYPE_LTE       -> true // ~ 10+ Mbps
			        // Unknown
			        //TelephonyManager.NETWORK_TYPE_UNKNOWN:
			        else -> false
		        }
	        else -> false
	    }
    }
}
